tests/robustness/validate/patch_history_test.go

```diff
@@ -0,0 +1,227 @@
+		},
+		{
+			name: "failed put remains if there is a matching event",
+			historyFunc: func(baseTime time.Time, h *model.AppendableHistory) {
+				start := time.Since(baseTime)
+				time.Sleep(time.Nanosecond)
+				stop := time.Since(baseTime)
+				h.AppendPut("key", "value", start, stop, nil, errors.New("failed"))
+			},
+			event: model.Event{
+				Type:  model.PutOperation,
+				Key:   "key",
+				Value: model.ToValueOrHash("value"),
+			},
+			expectRemains: true,
+		},
+		{
+			name: "failed put with lease remains if there is a matching event",
+			historyFunc: func(baseTime time.Time, h *model.AppendableHistory) {
+				start := time.Since(baseTime)
+				time.Sleep(time.Nanosecond)
+				stop := time.Since(baseTime)
+				h.AppendPutWithLease("key", "value", 123, start, stop, nil, errors.New("failed"))
+			},
+			event: model.Event{
+				Type:  model.PutOperation,
+				Key:   "key",
+				Value: model.ToValueOrHash("value"),
+			},
+			expectRemains: true,
+		},
+		{
+			name: "failed put is dropped",
+			historyFunc: func(baseTime time.Time, h *model.AppendableHistory) {
+				start := time.Since(baseTime)
+				time.Sleep(time.Nanosecond)
+				stop := time.Since(baseTime)
+				h.AppendPut("key", "value", start, stop, nil, errors.New("failed"))
+			},
+			expectRemains: false,
+		},
+		{
+			name: "failed put with lease is dropped",
+			historyFunc: func(baseTime time.Time, h *model.AppendableHistory) {
+				start := time.Since(baseTime)
+				time.Sleep(time.Nanosecond)
+				stop := time.Since(baseTime)
+				h.AppendPutWithLease("key", "value", 123, start, stop, nil, errors.New("failed"))
+			},
+			expectRemains: false,
+		},
+		{
+			name: "successful delete remains",
+			historyFunc: func(baseTime time.Time, h *model.AppendableHistory) {
+				start := time.Since(baseTime)
+				time.Sleep(time.Nanosecond)
+				stop := time.Since(baseTime)
+				h.AppendDelete("key", start, stop, &clientv3.DeleteResponse{}, nil)
+			},
+			expectRemains: true,
+		},
+		{
+			name: "failed delete remains",
+			historyFunc: func(baseTime time.Time, h *model.AppendableHistory) {
+				start := time.Since(baseTime)
+				time.Sleep(time.Nanosecond)
+				stop := time.Since(baseTime)
+				h.AppendDelete("key", start, stop, nil, errors.New("failed"))
+			},
+			expectRemains: true,
+		},
+		{
+			name: "successful txn remains",
+			historyFunc: func(baseTime time.Time, h *model.AppendableHistory) {
+				start := time.Since(baseTime)
+				time.Sleep(time.Nanosecond)
+				stop := time.Since(baseTime)
+				h.AppendTxn(nil, []clientv3.Op{}, []clientv3.Op{}, start, stop, &clientv3.TxnResponse{}, nil)
+			},
+			expectRemains: true,
+		},
+		{
+			name: "failed txn remains",
+			historyFunc: func(baseTime time.Time, h *model.AppendableHistory) {
+				start := time.Since(baseTime)
+				time.Sleep(time.Nanosecond)
+				stop := time.Since(baseTime)
+				h.AppendTxn(nil, []clientv3.Op{}, []clientv3.Op{}, start, stop, nil, errors.New("failed"))
+			},
+			expectRemains: false,
```

Comment:
Maybe rename the test case name to "non put/delete failed txn is dropped"?

Reply:
Doesn't match the previously established pattern. We could rename it to `failed empty txn remains`